### ISSUE TITLE
Contributing to Pytorch

### ISSUE THREAD

**Issue by yacth** [pmeier/pmeier#1]
Hi Philip,

I don't know if it's the right way to contact you for this kind of question but since you suggested it in your profile I am taking the shot.

I am willing to contribute to Pytorch to put on practice my knowledge in Python since I really enjoy the API, it could be fun to understand and work on what happens behind it.

Could you please guide me in some issues I could work on as a newbie.

Cheers,
Y.

**Comment by pmeier** [comment]
Hey @yacth 

> I don't know if it's the right way to contact you for this kind of question but since you suggested it in your profile I am taking the shot.

Yes this is the easiest way to contact me. Glad you reached out.

> I am willing to contribute to Pytorch to put on practice my knowledge in Python since I really enjoy the API, it could be fun to understand and work on what happens behind it.

That is awesome to hear! I'm not trying to discourage here, but PyTorch isn't really a beginner friendly project to contribute to. That is not that we don't welcome new contributors, but the source is just rather complex. Plus, PyTorch's core functionality are written in C++ and only wrapped in Python.

By looking at your profile I get that you haven't contributed to OSS before and thus don't know the typical workflow. Don't worry, everyone has to start somewhere. It just means you need to learn two skills at the same time, which can be overwhelming.

> Could you please guide me in some issues I could work on as a newbie.

In general you can look through the [issue tracker](https://github.com/pytorch/pytorch/issues) for the `good first issue` label. To filter the issues to only show ones labeled like that you can either select it in the "Label" drop down in the header bar or put `label:"good first issue"` in the search bar. Pick one that sounds interesting and read through it. Don't get discouraged if someone claimed they will work on it. Unfortunately they rarely follow through.

If you want to stay in "PyTorch land" it might also be useful to do the same for the domain libraries like [`torchvision`](https://github.com/pytorch/vision). They are mostly Python and thus should be easier to get started with.

One concrete issue that might be your starting point is pytorch/pytorch#59698. It is only fixing a link, but might get you started with the OSS workflow. Just comment on the issue if you want to take it up.

**Comment by yacth** [comment]
Hello @pmeier,

Thank you for your detailed answer.

Yes I didn't work much on OSS, I mostly programmed for my academic projects or my personal project.

I know a little about fork, branches and pull request.
I also understand C++ and some of the computer vision algorithms.

But I didn't find any issue that I could understand and contribute to for now.
Or maybe I don't know how to start.

If you have a suggestion on a project that I could participate too, I could start training on it.

**Comment by pmeier** [comment]
Lets have a look at the issue I pointed out above: pytorch/pytorch#59698. In there this file is reported as source of confusion:

https://github.com/pytorch/vision/blob/master/torchvision/models/utils.py

In there you have a common scheme:

```python
try:
    import foo
except ImportError:
    import bar as foo
```

We try to import `foo`, but if this fails, we import something similar an rename it to whatever we wanted to import in the first place. 

In the file above we try to import `torch.hub.load_state_dict_from_url` and in case that fails we use `torch.utils.model_zoo.load_url`. Since `torch.hub` is stable since almost two years, we don't need the fallback anymore. Plus, since `torch.utils.model_zoo` is deprecated in favor of `torch.hub`, the fallback will be invalid sooner or later.

To fix this, we can find every module that imports `load_state_dict_from_url` from `torchvision.models.utils` and simply replace it by an import from `torch.hub`. Afterwards we can delete `torchvision/models/utils.py`, because the file is no longer needed. 

Do you think you can start with this?

**Comment by yacth** [comment]
Hi Philip,

Thank you again for your detailed answer, I follow your instructions and give it a shot.
I will do a pull request when I will have my solution.

Small edit: 

Sorry I didn't advance much, I am stuck at the installation : https://github.com/pytorch/pytorch/blob/469f0e42d6e2b3cd8c78b224b97d45be2dc7d0ee/CONTRIBUTING.md#developing-pytorch 

The command `python setup.py install` gives me an error : 
```bash
Building wheel torch-1.10.0a0+git469f0e4
-- Building version 1.10.0a0+git469f0e4
cmake --build . --target install --config Release -- -j 8
make: Makefile: No such file or directory
make: *** No rule to make target 'Makefile'.  Stop.
```

I will look through it tomorrow, if you have any idea it would be great.

Thanks :).

**Comment by pmeier** [comment]
Hey Yacine, something like this is really hard to debug without having a look at your system. I did search for the error for a bit and it seems that you maybe don't have a compiler or an unsupported version. Could you run 

```shell
$ wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
$ python collect_env.py
```

and post the output here?

--- 

Apart from your build error, there are two more issues:

1. Installing PyTorch with only 8 threads is a lengthy endeavor. I would only advise to do this in case you want to work on the non-Python side of PyTorch. Since that seems not the case for you, have a look at the [nightly checkout tool](https://github.com/pytorch/pytorch/blob/469f0e42d6e2b3cd8c78b224b97d45be2dc7d0ee/CONTRIBUTING.md#nightly-checkout--pull) that gives you an up-to-date version of the compiled parts.
2. Although the issue was reported in PyTorch core, it is actually an issue for `torchvision`. I just transferred it to pytorch/vision#4069. This happens often for documentation issues, since there it is harder to see which repository that specific part belongs to. 

---

If you want to start developing on `torchvision` you also need an up-to-date version of `torch`, since both packages are developed in sync. Fortunately, you don't need to build `torch` here either, but you can just install a [nightly version](https://pytorch.org/get-started/locally/). Afterwards you can simply install `torchvision` with `python setup.py develop`.

Let me know if you need additional help. Please comment on the new issue in case you want to take it up so others know about it and we don't end up with duplicated work.

**Comment by pmeier** [comment]
Turns out this issue wasn't as easy to fix than I thought. Although PyTorch is mostly open source, there are some internal components that only FB employees can view. In this case `torchvision` needs the file, because they monkey-patch it to handle their internal infrastructure.

> If you have a suggestion on a project that I could participate too, I could start training on it.

There are so many projects that are worth contributing on that I could never get you an extensive list. If you want to stay in the same space PyTorch you could try

- [`numpy`](https://github.com/numpy/numpy),
- [`scipy`](https://github.com/scipy/scipy), or
- [`sklearn`](https://github.com/scikit-learn/scikit-learn). 

They all have the `good first issue label` and should be a lot easier to handle than PyTorch.

**Comment by yacth** [comment]
Ok thank you, I will definitely check that out.

I tried to follow their labeled `good first issue` but they are sometimes old or solved and it wasn't updated.

**Comment by pmeier** [comment]
> I tried to follow their labeled `good first issue` but they are sometimes old or solved and it wasn't updated.

Not all contributions have to be code. Quoting from the [`numpy` contribution guide](https://numpy.org/contribute/#issue-triaging):

> The [NumPy issue tracker](https://github.com/numpy/numpy/issues) has a *lot* of open issues. Some are no longer valid, some should be prioritized, and some would make good issues for new contributors. You can:
>
> - check if older bugs are still present
> -  find duplicate issues and link related ones
> - add good self-contained reproducers to issues
> - label issues correctly (this requires triage rights – just ask)
> 
> Please just dive in.

So if you have already checked that some of the issues are outdated, feel free to comment that on the issue. You should include an explanation for example linking a pull request. I'm sure the maintainers will be happy and although it may don't feel that way this is a valuable contribution.

---

I'm closing this issue now, because we no longer talk about contributing to PyTorch. I don't have any particular experience with projects listed above, but if you need some general guidance on the OSS workflow, feel free to open another issue. Have fun and good luck Yacine!

**Comment by pmeier** [comment]
Hey @yacth, I just wanted to check in real quick. How is your journey into the OSS world so far. I see you have a few PRs open at `scipy` which is awesome.

**Comment by yacth** [comment]
Hi @pmeier, thank you for getting back to me. Yes I am trying my best on contributing. I hope that I will be able to close them soon. I will try to find some other projects to work on after that.

**Comment by pmeier** [comment]
Glad to hear that. If you need some guidance, feel free to ping me on the PRs. Depending on the topic I might not be able to help with that, but I can certainly give suggestions on good pratices, design, and so on.

If you are looking for projects in the PyTorch ecosystem, I recently started to add more `good first issue`'s to one of my projects: [`pystiche`](https://pystiche.org).

**Comment by yacth** [comment]
Thank you @pmeier, sorry I wasn't available the past week.

I will definitely look at your project.
